Add support for Type=notify-reload with systemd

tie opened the pull request:
This PR adds support for using pgbouncer with `Type=notify-reload` systemd service.

See https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html

Note that `Type=notify-reload` services do not need `ExecReload=` command and use `ReloadSignal=` that conveniently defaults to SIGHUP.

eulerto commented:
I'm afraid we cannot accept this PR. This service type (`notify-reload`) was added in [version 253](https://github.com/systemd/systemd/releases/tag/v253). There are still supported operating systems that use old `systemd` versions like Ubuntu focal (245), Ubuntu jammy (249), Rocky 8 (239) and Rocky 9 (252). Once the main distributions support this feature we can merge this PR.

tie commented:
@eulerto, sorry, I don’t understand your point. This PR makes a single change that could be incompatible with older systemd versions, that is, updates Type= in `etc/pgbouncer.service`. If some distributions use the provided example unit as-is, we can document that pgbouncer supports `Type=notify-reload` in the comment above instead and update it later. `sd_notifyf` has been available for decades and `Type=notify` should simply ignore unknown fields like `MONOTONIC_USEC`.

tie commented:
> etc/pgbouncer.service

And from what I understand, this file is used for reference/documentation purposes only. That is, it is installed under `docdir` and other distributions ship their own service definitions, e.g.
- Debian: https://salsa.debian.org/postgresql/pgbouncer/-/blob/460febb15ab9f040e74da9b2d4023fe3bb5c8481/debian/pgbouncer.service
- (Ubuntu package is mostly derived from Debian)
- Ubuntu (focal): https://git.launchpad.net/ubuntu/+source/pgbouncer/tree/debian?h=applied/ubuntu/focal
- Ubuntu (jammy): https://git.launchpad.net/ubuntu/+source/pgbouncer/tree/debian?h=applied/ubuntu/jammy
- Not sure where Rocky Linux code is hosted.

JelteF commented:
I think the `sd_notifyf` call should be fine to merge. Assuming that it indeed does not break older systemd setups.

But I do agree with @eulerto and the default systemd config which people copy paste should work with at least the majority of systems that people are expected to run in production environments. Since both the latest RHEL and latest Ubuntu LTS don't even support this feature, currently is not the time for that.

I'd be okay with merging a change to the comments of the config file, mentioning what could be change for systemd 253+. So basically inverting the comment addition that you made.

JelteF commented:
To be clear, I do agree that most people will use the systemd file from their package management. But even that being the case, showing a systemd file in the docs that is broken for pretty much all systems that we expect people to run at the moment seems like a bad choice.

tie commented:
@JelteF, rebased and inverted the comment. I’ve also noticed that I’ve wrongly assumed that `get_cached_time`/`get_time_usec` from libusual uses `CLOCK_MONOTONIC` that is required for proper reload synchronization with systemd (currently [`get_time_usec`](https://github.com/libusual/libusual/blob/f24c5ea6abfb549dfec09c3bb2a3f1fbb99faaa6/usual/time.c#L68-L74) uses [`gettimeofday`](https://blog.habets.se/2010/09/gettimeofday-should-never-be-used-to-measure-time.html). I’ve added explicit `clock_gettime` call in this PR.

eulerto commented on src/main.c:
```diff
@@ -565,10 +565,31 @@ static void handle_sigusr2(int sock, short flags, void *arg)
+	usec_t usec;
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	usec = (usec_t)ts.tv_sec * USEC + (usec_t)ts.tv_nsec / (usec_t)1000;
+	sd_notifyf(0, "RELOADING=1\nMONOTONIC_USEC=%" PRIu64, usec);
```
Does systemd ignore the MONOTONIC_USEC if the version is less than 253? AFAICS, yes. However, I didn't test.

tie replied:
Yes, at least systemd 252 seems to be fine with this.

eulerto reviewed (commented):
My main objection is related to the `pgbouncer.service` modification. It is just a comment, I'm fine.

eulerto reviewed (approved):
LGTM
